lib/tpm2_capability.c

```diff
@@ -40,118 +40,123 @@
 
         if (rval != TPM2_RC_SUCCESS) {
             LOG_ERR("Failed to GetCapability: capability: 0x%x, property: 0x%x",
                      capability, property);
             LOG_PERR(ESys_GetCapability, rval);
+            if (*capability_data) {
+                free(*capability_data);
+                *capability_data = NULL;
+            }
             return false;
         }
 
-        /* allocate memory for consolidated capabilities list */
-        TPMS_CAPABILITY_DATA *updated = calloc(1, sizeof(TPMS_CAPABILITY_DATA));
-        if (!updated) {
-            LOG_ERR("oom");
+        if (fetched_data->capability != capability) {
+            LOG_ERR("TPM returned different capability than requested: 0x%x != 0x%x",
+                     fetched_data->capability, capability);
             free(fetched_data);
+            if (*capability_data) {
+                free(*capability_data);
+                *capability_data = NULL;
+            }
             return false;
         }
 
-        UINT32 i;
-        UINT32 existing = 0;
-        /* Copy all existing results into the newly allocated area */
-        if (*capability_data) {
-            existing = (*capability_data)->data.algorithms.count;
-            for (i = 0; i < existing; i++) {
-                updated->data.algorithms.algProperties[i] =
-                    (*capability_data)->data.algorithms.algProperties[i];
-            }
-            updated->data.algorithms.count = existing;
+        if (*capability_data == NULL) {
+            /* reuse the TPM's result structure */
+            *capability_data = fetched_data;
 
-            free(*capability_data);
-            *capability_data = NULL;
+            if (!more_data) {
+                /* there won't be another iteration of the loop, just return the result unmodified */
+                return true;
+            }
         }
 
-        /* copy the newly retrieved data in too */
-        UINT32 new_items = fetched_data->data.algorithms.count;
-        for (i = 0; i < new_items; i++) {
-            updated->data.algorithms.algProperties[i] = fetched_data->data.algorithms.algProperties[i];
-            updated->data.algorithms.count++;
+        /* append the TPM's results to the initial structure, as long as there is still space left */
+        switch (capability) {
+            case TPM2_CAP_ALGS:
+                APPEND_CAPABILITY_INFORMATION(algorithms, algProperties, .alg, TPM2_MAX_CAP_ALGS);
+                break;
+            case TPM2_CAP_HANDLES:
+                APPEND_CAPABILITY_INFORMATION(handles, handle, , TPM2_MAX_CAP_HANDLES);
+                break;
+            case TPM2_CAP_COMMANDS:
+                APPEND_CAPABILITY_INFORMATION(command, commandAttributes, , TPM2_MAX_CAP_CC);
+                /* workaround because tpm2-tss does not implement attribute commandIndex for TPMA_CC */
+                property &= TPMA_CC_COMMANDINDEX_MASK;
```

Comment:
Is this a bug in tpm2-tss? If so can we file it there and reference in the comments so that we can remove this workaround at some point?